### ISSUE TITLE
What RaptorJIT fixes

### ISSUE THREAD

**Issue by lukego** [raptorjit/raptorjit#213]
I have been thinking about the best way to explain the purpose of the RaptorJIT project and especially how I perceive the goals of this project to differ from LuaJIT. Today I remembered the idea of writing down short sentences summarizing [What Languages Fix](http://www.paulgraham.com/fix.html).

Here's an attempt in the context of RaptorJIT. These are necessarily extreme oversimplifications but I hope they are useful nevertheless.

- C: I want to write an application with high performance.
- Lua: I want a much easier way to write the parts that don't have to run so fast.
- LuaJIT: I have lots of Lua code and now I want to make it run twice as fast (at least.)
- RaptorJIT: I want to use Lua instead of C for the high-performance parts too.

So RaptorJIT is competing with C. Users need to be able to write efficient code with confidence. This requires the compiler algorithms to be simple and predictable. It also requires convenient profiling tools to make the actual runtime behaviour of applications transparent and quick to understand.

LuaJIT is competing with Lua. Users have already written their applications using PUC Lua and now they want them to run faster. This requires the compiler to be "magical" and to automatically speed up existing Lua code written by people who didn't have the JIT in mind. There's no need to make the compiler simple because users aren't going to learn how it works anyway. The most important thing is to have a set of heuristics that tends to match the Lua code found in the wild. And to heavily optimize the fallback interpreter to "win" even on code that's fundamentally mismatched to the JIT.

So RaptorJIT is solving a different problem than LuaJIT and, I dare say, an easier one. The really hard problem is to take an arbitrary existing application and make it run really fast. (That's the hill that [TraceMonkey](http://www.cs.williams.edu/~freund/cs434/gal-trace.pdf) died on.) RaptorJIT "only" has to help users to confidently take responsibility for the performance of their applications themselves.

**Comment by szbnwer** [comment]
i can confirm that this is kinda much foolproof! :D

otoh, that link seems to be promising for getting nearer to enlightenment, thx for that! :)

**Comment by dormando** [comment]
Think I've self-answered all the questions I have about raptorJIT by trawling through these issues and PR's... but maybe it'd be a waste to not actually ask anything. Hijacking this issue since the answer here seems closest to answer my original question.

- I need an embedded language that's easy to use, but performance close to C for small subsets of functionality. Does this language fit?

Lets ask a bunch more questions to answer this big one.

- How you you embed raptorJIT into a C app?

This isn't addressed anywhere I can find. I downloaded and built everything; it's all quite nice, but if I wanted to find out how to embed it I would have to start from a luajit specific guide. Further the license is Apache 2.0 mixed with BSD/MIT, so some more care has to be taken if shipping the VM source inside another codebase (which is my personal intention) It's not bad I just had to dig around here.

- Do the maintainers of the project intend it to be an embedded interpreter, standalone, or both?

The nature of the README, this issue, etc, sound like standalone is the ultimate goal. Even if embedding works now, I would worry that down the line it becomes harder or impossible without losing track of upstream.

- How is development done?

Man I love this part. "The linux kernel model" is how I do everything. Scary things sit in a branch until they're more or less proven. I see the VM ASMtoC branch and a lot of text around wanting the portability and where the intention is. Given that I have faith that if the C VM were a catastrophic performance problem, it would start to violate https://github.com/lukego/blog/issues/21 as a standard, and either take more time or have adjusted goals.

Compilers have gotten a bit better and computers have gotten better at running bad code, so the C VM doesn't seem too scary. My only problem here is these issues all talk of the upside of a C VM, but aren't clear about what's being lost. That seems to just not be known until the C work is actually done though.

- How are users intended to use the language?

This goal (and the text around the VM rewrite) seems to answer this: as a deliberate application language where the authors are in control of optimization. For me, that's C. The lua will be written mostly by end users, using examples/libraries that I write. I don't get any benefit from rewriting the rest in lua.

The idea here is that "non-JIT'able code should be an error" and the tooling (studio, profiling, etc) expands to fill the gap. This doesn't absolutely mean I can't use it, but it means that I may need to build in tooling like:
1. internal counters to inform users when the JIT is being skipped too much
2. end user analysis tools and examples for writing well-optimized code
3. preprocessing or boot-time testing to make code which would perform poorly an error. I'm probably not smart enough to accomplish this.

Varnish Control Language (VCL) is a good example of where you land when taking this line of thought. Ignoring the implementation (a VCL -> C transpiler written in TCL), it's a C-like language that's cut down to make the default modes fast, enough that it _looks to the user_ like they're going too far if they build things that won't work well. A VCL-alike would actually work for me, if users would tolerate having gcc as a runtime dependency. Though you end up losing a lot of flexibility with this route.

---

Think everything else was trivial to answer. It's a fork of LuaJIT, after all.

In short; I have a surface area of configuration-style code that doesn't need to be fast (PUC lua would be fine). Then there's a layer of usually simple branch logic that if I could use lua coroutines/fibers (or short-circuit to make such even more efficient when state isn't necessary) would be extremely powerful for the program. Lua and VCL, more or less.

**Comment by lukego** [comment]
> How you you embed raptorJIT into a C app?

Just answering this directly now. It's not covered in the README but we do this in Snabb and the Makefile is here: https://github.com/snabbco/snabb/blob/master/src/Makefile

Our basic approach is:

1. Have a `main.c` that starts the VM and runs the code found in `startup.lua`.
2. Statically link `libraptorjit.a` into the executable.
3. Statically link each Lua module into the executable by first converting it to `.o`. Our Makefile automatically detects all the `*.lua` modules and compiles them into the binary so added a new one doesn't require a Makefile change.
4. Convert all `*.h` header files into wrapper Lua modules that can be `require()`d to load the definitions into the FFI.
5. Compile and link all `*.c` code too since we occasionally write some of that.

(We also have some extra steps e.g. preprocessing Lua modules that generate machine code with DynASM but that's a bit more exotic.)

So what you get is an executable that is fairly self-contained and only depends on libc to run. You can still load Lua source files at runtime etc if that is what you want to do but we usually don't since we can build a new binary in a couple of seconds and it only takes up a couple of megabytes.

Maybe we should make a project template like this so that it's easy to adopt for other people who want to do the same?

**Comment by lukego** [comment]
> Compilers have gotten a bit better and computers have gotten better at running bad code, so the C VM doesn't seem too scary. My only problem here is these issues all talk of the upside of a C VM, but aren't clear about what's being lost. That seems to just not be known until the C work is actually done though.

True. I am expecting the C VM to have performance competitive with the assembler VM, but that is still speculation at this stage. It's possible that it will be slower and that this might make it hard to merge.

However, you can also profile an application to estimate how much time it spends in the bytecode interpreter. My experience has been that it's usually around 0.0% of time. If the application is only spending a few milliseconds during startup running the interpreter and the JIT compiler then it really limits the potential benefit of optimizing those components.

> The idea here is that "non-JIT'able code should be an error" and the tooling (studio, profiling, etc) expands to fill the gap. This doesn't absolutely mean I can't use it, but it means that I may need to build in tooling like:

I take your point.

I think that RaptorJIT makes this tooling quite straightforward to provide because the profiler has no overhead. You could use standard profiler data to automatically report "good/okay/bad" optimization of the Lua code - these could be standard metrics counted in the VM at all times. That's just comparing the ratios of time spent in various VM states e.g. interpreted verses compiled verses garbage collector.

That's a bit like the basic diagnostic of using `vmstat` to check if your computer is running the CPU at 100%, or swapping like mad, etc.
